src/hw/pcie_cfg.cpp

```diff
@@ -0,0 +1,194 @@
+#include "options.hpp"
+
+#include <phosphor-logging/log.hpp>
+#include <xyz/openbmc_project/Control/PCIe/server.hpp>
+
+#include <charconv>
+
+using namespace phosphor::logging;
+
+using namespace sdbusplus::xyz::openbmc_project::Control::server;
+using BifurcationConfiguration =
+    std::vector<std::tuple<uint8_t, uint8_t, PCIe::BifurcationMode>>;
+
+// PCIe Bifurcation mode
+static constexpr uint8_t pcieBifurcateX4X4X4X4 = 0;
+static constexpr uint8_t pcieBifurcateX4X4XXX8 = 1;
+static constexpr uint8_t pcieBifurcateXXX8X4X4 = 2;
+static constexpr uint8_t pcieBifurcateXXX8XXX8 = 3;
+static constexpr uint8_t pcieBifurcateXXXXXX16 = 4;
+static constexpr uint8_t pcieBifurcateXXXXXXXX = 0xF;
+
+/**
+ * @brief Lookup dbus service for interface
+ *
+ * @param[in] interface     DBus interface to lookup
+ * @return service name and resource path
+ */
+static std::tuple<std::string, std::string>
+    dbusGetSetviceAndPath(sdbusplus::bus::bus& bus,
+                          const std::string& interface)
+{
+    auto getObjects =
+        bus.new_method_call(dbus::mapper::busName, dbus::mapper::path,
+                            dbus::mapper::interface, dbus::mapper::subtree);
+    getObjects.append("/", 0, std::array<std::string, 1>{interface});
+
+    SubTreeType objects;
+    bus.call(getObjects).read(objects);
+
+    if (objects.size() != 1)
+    {
+        throw sdbusplus::exception::SdBusError(-EINVAL,
+                                               "unexpected objects count");
+    }
+    if (objects.begin()->second.size() != 1)
+    {
+        throw sdbusplus::exception::SdBusError(-EINVAL,
+                                               "unexpected services count");
+    }
+
+    auto path = objects.begin()->first;
+    auto service = objects.begin()->second.begin()->first;
+    return std::make_tuple(path, service);
+}
+
+bool pcieCfg::addBifurcationConfig(const int& socket,
+                                   const std::string& optValue)
+{
+    uint16_t instance = 0;
+    uint8_t value = 0;
+    static const std::regex optionRegex(
+        "[a-f0-9]{4}", std::regex::icase | std::regex::optimize);
+    if (!std::regex_match(optValue, optionRegex))
+    {
+        log<level::ERR>("Invalid PCIe configuration option format",
+                        entry("VALUE=%s", optValue.c_str()));
+        return false;
+    }
+
+    std::from_chars(optValue.data() + 0, optValue.data() + 2, instance, 16);
+    std::from_chars(optValue.data() + 2, optValue.data() + 4, value, 16);
+    instance |= (socket & 0xFF) << 8;
+
+    // check if other component already claim to use this port and try to merge
+    // configs for port halves We assume that port can be only shared between
+    // components by halves (8 lanes) and for not used half the component always
+    // claim 'x8' mode.
+    auto cfg = bifurcationConfig.find(instance);
+    if (cfg != bifurcationConfig.end())
+    {
+        // get configuration, requested by other consumer
```

Comment:
There is no consumer here. Also, the comma is not needed in English before a participial.
Let's say "Get the config previously set for the port by another hardware component".